### PR TITLE
Queues - Erica Case - Ride share two

### PR DESCRIPTION
**Pull request by EricaJCasePhD**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Probably the decisions I agonized over the most were whether to raise and rescue errors or output a warning. Coming up with a conceptual framework helped me make final decisions: errors and rescues for bad data (ie driver_id's, vins, etc that did not make sense) and warnings for bad user queries or data that was missing due to a logical explanation (ie riders who signed up for ride-share but haven't used it yet). |
| Describe a concept that you gained more clarity on as you worked on this assignment.|I gained clarity on how to handle bad data by either removing it from the dataset or letting the program take its default action (ie returning an empty array). |
| Describe a nominal test that you wrote for this assignment.|For pretty much every csv, I tested whether my code correctly converted the first and last rows of a csv to an object (ie the riders csv to an array of class Rider objects. |
| Describe an edge case test that you wrote for this assignment.|The case: a rider id is included in the trips file that is not present in the rider file (ie rider #0). The desired result: the trips with rider 0 should not be included in the trips array via argument error and rescue. The test: `trips.all(rider_0).must_raise Argument Error` all trips will not have any trips with rider id equal to 0
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Uh, I still wish I'd done that. I was trying to save time and catch up, but ended up wasting a lot of time with unfocused coding instead. Where are you, time machine? |

### CODE REVIEW

File: lib/records.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    def self.all(search_var)
+      csv = CSV.read(self.csv_name,
+      { headers: true,
+        header_converters: :symbol,
+        converters: :all  # format Integers
+        })
+
+
+        csv.inject([])  do |arr, row|
```
Comment (droberts-sea) [inline lib/records.rb:24]: Good use of enumerables here.

File: lib/records.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    def self.add_record(args, search_var)
+      self.new(args, search_var)
+    rescue ArgumentError => e
+      puts "#{e.message}"
+    rescue VinError => e
+      puts "#{e.message}"
+    end
+
+    def self.all(search_var)
```
Comment (droberts-sea) [inline lib/records.rb:16]: This `search_var` variable ends up being a little clunky - you have to pass it as an extra variable through all those method calls.

A different way to approach it is to assume that anything inheriting from `Record` implement some method, call it `id_field`, which returns the name of the ID field as a string (essentially what `search_var` is tracking). Then in `Record`, when you're loading data you just call that method whenever you need it rather than lugging a variable around.

Of course, that will break your implementation of `find`. You might be interested in the `send` method, which calls a method by name. In context:

```ruby
def find(field, value)
  all.select do |record|
    record.send(field) == value
  end
end
```

File: lib/records.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
+module Rideshare
+  class Record
```
Comment (droberts-sea) [inline lib/records.rb:2]: Good work abstracting away some of the repeated functionality! I'm glad you were able to get this working.

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+            it "correctly calculates the average rating for the first driver based on all trips taken" do
+              first_driver_rating.must_equal 2.6
+            end
+
+            it "correctly calculates the average rating for the last driver based on all trips taken" do
+              last_driver_rating.must_equal 2.8
+            end
+
+            it "returns nil and outputs a warning if the driver has not done any properly rated trips yet" do
```
Comment (droberts-sea) [inline specs/driver_spec.rb:167]: Good test case! Edge cases like this, asking what happens when there *isn't* any data, can be easy to miss.
